### PR TITLE
Common flags across VPA components

### LINKED ISSUE #7291: VPA: Passing args to vpa-updater
The moment I pass any arg to updater pod ex: as described [here](https://github.com/kubernetes/autoscaler/blob/master/vertical-pod-autoscaler/FAQ.md#i-get-recommendations-for-my-single-pod-replicaset-but-they-are-not-applied)  , I stop getting any logs for updater after the logs as shown below. What am I doing wrong ?  
```
I0917 08:33:45.292806       1 fetcher.go:99] Initial sync of DaemonSet completed
I0917 08:33:45.493064       1 fetcher.go:99] Initial sync of Deployment completed
I0917 08:33:45.593691       1 fetcher.go:99] Initial sync of ReplicaSet completed
I0917 08:33:45.694550       1 fetcher.go:99] Initial sync of StatefulSet completed
I0917 08:33:45.794948       1 fetcher.go:99] Initial sync of ReplicationController completed
I0917 08:33:45.895580       1 fetcher.go:99] Initial sync of Job completed
I0917 08:33:45.996170       1 fetcher.go:99] Initial sync of CronJob completed
I0917 08:33:45.996292       1 controller_fetcher.go:141] Initial sync of Deployment completed
I0917 08:33:45.996313       1 controller_fetcher.go:141] Initial sync of ReplicaSet completed
I0917 08:33:45.996318       1 controller_fetcher.go:141] Initial sync of StatefulSet completed
I0917 08:33:45.996322       1 controller_fetcher.go:141] Initial sync of ReplicationController completed
I0917 08:33:45.996326       1 controller_fetcher.go:141] Initial sync of Job completed
I0917 08:33:45.996329       1 controller_fetcher.go:141] Initial sync of CronJob completed
I0917 08:33:45.996333       1 controller_fetcher.go:141] Initial sync of DaemonSet completed
W0917 08:33:45.996380       1 shared_informer.go:459] The sharedIndexInformer has started, run more than once is not allowed
W0917 08:33:45.996377       1 shared_informer.go:459] The sharedIndexInformer has started, run more than once is not allowed
W0917 08:33:45.996397       1 shared_informer.go:459] The sharedIndexInformer has started, run more than once is not allowed
W0917 08:33:45.996407       1 shared_informer.go:459] The sharedIndexInformer has started, run more than once is not allowed
W0917 08:33:45.996412       1 shared_informer.go:459] The sharedIndexInformer has started, run more than once is not allowed
W0917 08:33:45.996416       1 shared_informer.go:459] The sharedIndexInformer has started, run more than once is not allowed
W0917 08:33:45.996379       1 shared_informer.go:459] The sharedIndexInformer has started, run more than once is not allowed
I0917 08:33:46.598356       1 api.go:94] Initial VPA synced successfully
``` 
Note: I manually changes the min-replica default value in source and built the docker image. It worked fine.

### PR DESCRIPTION
**Pull request by omerap12**
#### What type of PR is this?
/kind cleanup

#### What this PR does / why we need it:
context: https://github.com/kubernetes/autoscaler/issues/7291#issuecomment-2373848344
We aim to introduce a common flags package to be used across all VPA components (updater, recommender, and admission-controller), consolidating shared configuration options. Currently, arguments are passed using CMD in the Dockerfile, which can lead to improper behavior when users specify custom arguments. By removing this and setting default values under containers.x.args, we eliminate the problematic behavior described above.

#### Which issue(s) this PR fixes:

Fixes #7291

#### Special notes for your reviewer:
If approved, we will need to update the VPA chart to include the default arguments (currently specified in the Dockerfile).
#### Does this PR introduce a user-facing change?

```release-note
None
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
None
```

### CODE REVIEW

**Comment by omerap12** [comment]
cc @adrianmoisey

**Comment by adrianmoisey** [comment]
> If approved, we will need to update the VPA chart to include the default arguments (currently specified in the Dockerfile).

Is there a way to move this parameter into code?

File: vertical-pod-autoscaler/pkg/recommender/Dockerfile [hunk 1]
```diff
@@ -31,5 +31,4 @@ ARG TARGETARCH
 
 COPY --from=builder /gopath/src/k8s.io/autoscaler/vertical-pod-autoscaler/pkg/recommender/recommender-$TARGETARCH /recommender
 
-ENTRYPOINT ["/recommender"]
-CMD ["--v=4", "--stderrthreshold=info", "--prometheus-address=http://prometheus.monitoring.svc"]
```
Comment (adrianmoisey) [inline vertical-pod-autoscaler/pkg/recommender/Dockerfile:35]: Should this be moved into the recommender as a default value?

**Comment by omerap12** [comment]
> > If approved, we will need to update the VPA chart to include the default arguments (currently specified in the Dockerfile).
> 
> Is there a way to move this parameter into code?

AFAIK no, because those parameter are configured inside the klog package

File: vertical-pod-autoscaler/pkg/recommender/Dockerfile (hunk 1 above)
Comment (omerap12) [reply vertical-pod-autoscaler/pkg/recommender/Dockerfile:35]: I accidentally deleted the ENTRYPOINT as well ( will fix ) , the prometheus address will be configured as default here: https://github.com/kubernetes/autoscaler/pull/7312/files#diff-cf801d600c83feda2a9ee187a09a43380478c36df274f42f2246f2de2a203505R48
You are correct, I should changed its name to  `prometheus-address`

File: vertical-pod-autoscaler/pkg/updater/Dockerfile [hunk 2]
```diff
@@ -31,5 +31,4 @@ ARG TARGETARCH
 
 COPY --from=builder /gopath/src/k8s.io/autoscaler/vertical-pod-autoscaler/pkg/updater/updater-$TARGETARCH /updater
 
-ENTRYPOINT ["/updater"]
-CMD ["--v=4", "--stderrthreshold=info"]
+ENTRYPOINT ["/updater"]
```
[the remark this answers is not in the document]
Comment (omerap12) [reply vertical-pod-autoscaler/pkg/updater/Dockerfile:34]: Of course, will fix

File: vertical-pod-autoscaler/common/flags.go [hunk 3]
```diff
@@ -0,0 +1,54 @@
+/*
+Copyright 2018 The Kubernetes Authors.
```
[the remark this answers is not in the document]
Comment (omerap12) [reply vertical-pod-autoscaler/common/flags.go:2]: Yeap

File: vertical-pod-autoscaler/pkg/admission-controller/Dockerfile [hunk 4]
```diff
@@ -32,4 +32,3 @@ ARG TARGETARCH
 COPY --from=builder /gopath/src/k8s.io/autoscaler/vertical-pod-autoscaler/pkg/admission-controller/admission-controller-$TARGETARCH /admission-controller
 
 ENTRYPOINT ["/admission-controller"]
-CMD ["--v=4", "--stderrthreshold=info"]
```
Comment (jackfrancis) [inline vertical-pod-autoscaler/pkg/admission-controller/Dockerfile:35]: It seems that adding flag variable re-use in a common place and cleaning up the static arguments in the various Dockerfiles are separate changes. Should there be a separate PR for each?

Comment (omerap12) [reply vertical-pod-autoscaler/pkg/admission-controller/Dockerfile:35]: I can separate this to two PRs if you prefer I don't mind

File: vertical-pod-autoscaler/pkg/recommender/Dockerfile (hunk 1 above)
Comment (adrianmoisey) [reply vertical-pod-autoscaler/pkg/recommender/Dockerfile:35]: That seems wrong. `address` and `prometheus-address` are two different parameters

Comment (adrianmoisey) [reply vertical-pod-autoscaler/pkg/recommender/Dockerfile:35]: `prometheus-address` is listed here: https://github.com/kubernetes/autoscaler/pull/7312/files#diff-bc201689c0c988297d4f93cf844937aa4f4755bfdc5bc912746148eaeed3bb64R61
And it has no default

Comment (omerap12) [reply vertical-pod-autoscaler/pkg/recommender/Dockerfile:35]: Oh yes I see sorry ( got confused ) . I guess I can move set it too as a default value.

**Comment by adrianmoisey** [comment]
> > > If approved, we will need to update the VPA chart to include the default arguments (currently specified in the Dockerfile).
> > 
> > 
> > Is there a way to move this parameter into code?
> 
> AFAIK no, because those parameter are configured inside the klog package

It seems they're possible to override, something like this seems to work:
```
	klog.InitFlags(nil)
	flag.Lookup("v").Usage = "set the log level verbosity (default: 2)"
	_ = flag.Set("v", "2")
	flag.Parse()
```

File: vertical-pod-autoscaler/pkg/admission-controller/Dockerfile (hunk 4 above)
Comment (jackfrancis) [reply vertical-pod-autoscaler/pkg/admission-controller/Dockerfile:35]: I think that makes the most sense, thanks!

File: vertical-pod-autoscaler/pkg/updater/main.go [hunk 5]
```diff
@@ -51,9 +50,6 @@ import (
 )
 
 var (
-	updaterInterval = flag.Duration("updater-interval", 1*time.Minute,
-		`How often updater should run`)
```
Comment (adrianmoisey) [inline vertical-pod-autoscaler/pkg/updater/main.go:55]: This one is being removed, should it be?

Comment (adrianmoisey) [reply vertical-pod-autoscaler/pkg/updater/main.go:55]: Oh, I see it was changed for `--interval`
My vote is to not change any parameters, just move the common ones

File: vertical-pod-autoscaler/pkg/updater/main.go [hunk 6]
```diff
@@ -66,18 +62,10 @@ var (
 
 	evictionRateBurst = flag.Int("eviction-rate-burst", 1, `Burst of pods that can be evicted.`)
 
-	address         = flag.String("address", ":8943", "The address to expose Prometheus metrics.")
```
Comment (adrianmoisey) [inline vertical-pod-autoscaler/pkg/updater/main.go:69]: Address is removed from each component and added to the common flags, but each component has a different default. 
I think this behaviour shouldn't change.

File: vertical-pod-autoscaler/pkg/admission-controller/Dockerfile (hunk 4 above)
Comment (adrianmoisey) [reply vertical-pod-autoscaler/pkg/admission-controller/Dockerfile]: Given that the point of this PR is to move defaults out of the Dockerfile into code, does it make sense to re-add the Dockerfile change, making the PR more atomic?

File: vertical-pod-autoscaler/pkg/updater/main.go (hunk 5 above)
Comment (omerap12) [reply vertical-pod-autoscaler/pkg/updater/main.go:55]: Sure, will be fixed

File: vertical-pod-autoscaler/pkg/updater/main.go (hunk 6 above)
Comment (omerap12) [reply vertical-pod-autoscaler/pkg/updater/main.go:69]: Will be fixed too.

File: vertical-pod-autoscaler/pkg/updater/main.go [hunk 7]
```diff
@@ -51,12 +50,12 @@ import (
...
 var (
-	updaterInterval = flag.Duration("updater-interval", 1*time.Minute,
-		`How often updater should run`)
-
 	minReplicas = flag.Int("min-replicas", 2,
 		`Minimum number of replicas to perform update`)
 
+	updaterInterval = flag.Duration("updater-interval", 1*time.Minute,
+		`How often updater should run`)
+
```
Comment (adrianmoisey) [inline vertical-pod-autoscaler/pkg/updater/main.go:58]: Super nitpick, sorry about that.
Can this be put back where it was?
Just to keep git history clean.

File: vertical-pod-autoscaler/pkg/admission-controller/Dockerfile (hunk 4 above)
Comment (adrianmoisey) [reply vertical-pod-autoscaler/pkg/admission-controller/Dockerfile:35]: @jackfrancis just want to see if you're fine with the Dockerfile change happening in this PR now. This PR is basically moving defaults out of the Dockerfile into the code.

File: vertical-pod-autoscaler/pkg/updater/main.go (hunk 7 above)
Comment (omerap12) [reply vertical-pod-autoscaler/pkg/updater/main.go:58]: yes, of course

Comment (omerap12) [reply vertical-pod-autoscaler/pkg/updater/main.go]: Done

File: vertical-pod-autoscaler/common/flags.go [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+	flag.StringVar(&cf.VpaObjectNamespace, "vpa-object-namespace", apiv1.NamespaceAll, "Namespace to search for VPA objects. Empty means all namespaces will be used.")
+	flag.StringVar(&cf.IgnoredVpaObjectNamespaces, "ignored-vpa-object-namespaces", "", "Comma separated list of namespaces to ignore when searching for VPA objects. Empty means no namespaces will be ignored.")
+	return cf
+}
+
+// InitLoggingFlags initializes the logging flags
+func InitLoggingFlags() {
+	flag.Lookup("v").Usage = "set the log level verbosity (default: 2)"
+	flag.Lookup("stderrthreshold").Usage = "set the log level threshold for writing to standard error (default: info)"
+	err := flag.Set("v", "2")
```
Comment (adrianmoisey) [inline vertical-pod-autoscaler/common/flags.go:54]: The Dockerfile ENTRYPOINT used to set this as a 4, I think it makes sense to set 4 as default, in order to not change behaviour

File: vertical-pod-autoscaler/pkg/admission-controller/main.go [hunk 9]
```diff
@@ -66,41 +65,38 @@ var (
...
+	webHookFailurePolicy = flag.Bool("webhook-failure-policy-fail", false, "If set to true, will configure the admission webhook failurePolicy to \"Fail\". Use with caution.")
+	registerWebhook      = flag.Bool("register-webhook", true, "If set to true, admission webhook object will be created on start up to register with the API server.")
+	registerByURL        = flag.Bool("register-by-url", false, "If set to true, admission webhook will be registered by URL (webhookAddress:webhookPort) instead of by service name")
 )
 
 func main() {
+	commonFlags := common.InitCommonFlags()
 	klog.InitFlags(nil)
+	common.InitLoggingFlags()
+	flag.Parse()
```
Comment (adrianmoisey) [inline vertical-pod-autoscaler/pkg/admission-controller/main.go:84]: Is this needed?

Comment (adrianmoisey) [reply vertical-pod-autoscaler/pkg/admission-controller/main.go:84]: It seems to change the `--help` output

Comment (omerap12) [reply vertical-pod-autoscaler/pkg/admission-controller/main.go]: The function `common.InitLoggingFlags()` is used across all VPA components to avoid code duplication.

File: vertical-pod-autoscaler/common/flags.go (hunk 8 above)
Comment (omerap12) [reply vertical-pod-autoscaler/common/flags.go]: Fixed

**Comment by adrianmoisey** [comment]
/assign @jbartosik 
Would you have time to review this?

File: vertical-pod-autoscaler/common/flags.go [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+	flag.Float64Var(&cf.KubeApiBurst, "kube-api-burst", 10.0, "QPS burst limit when making requests to Kubernetes apiserver")
+	flag.BoolVar(&cf.EnableProfiling, "profiling", false, "Is debug/pprof endpoint enabled")
+	flag.StringVar(&cf.VpaObjectNamespace, "vpa-object-namespace", apiv1.NamespaceAll, "Namespace to search for VPA objects. Empty means all namespaces will be used.")
+	flag.StringVar(&cf.IgnoredVpaObjectNamespaces, "ignored-vpa-object-namespaces", "", "Comma separated list of namespaces to ignore when searching for VPA objects. Empty means no namespaces will be ignored.")
+	return cf
+}
+
+// InitLoggingFlags initializes the logging flags
+func InitLoggingFlags() {
+	flag.Lookup("v").Usage = "set the log level verbosity (default: 2)"
```
Comment (jackfrancis) [inline vertical-pod-autoscaler/common/flags.go:52]: These too `Lookup()` invocations are subject to a nil pointer panic:

https://pkg.go.dev/flag#Lookup

File: vertical-pod-autoscaler/common/flags.go [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+	flag.Float64Var(&cf.KubeApiQps, "kube-api-qps", 5.0, "QPS limit when making requests to Kubernetes apiserver")
+	flag.Float64Var(&cf.KubeApiBurst, "kube-api-burst", 10.0, "QPS burst limit when making requests to Kubernetes apiserver")
+	flag.BoolVar(&cf.EnableProfiling, "profiling", false, "Is debug/pprof endpoint enabled")
+	flag.StringVar(&cf.VpaObjectNamespace, "vpa-object-namespace", apiv1.NamespaceAll, "Namespace to search for VPA objects. Empty means all namespaces will be used.")
+	flag.StringVar(&cf.IgnoredVpaObjectNamespaces, "ignored-vpa-object-namespaces", "", "Comma separated list of namespaces to ignore when searching for VPA objects. Empty means no namespaces will be ignored.")
+	return cf
+}
+
+// InitLoggingFlags initializes the logging flags
+func InitLoggingFlags() {
```
Comment (jackfrancis) [inline vertical-pod-autoscaler/common/flags.go:51]: Because we are always calling `InitLoggingFlags` with `InitCommonFlags`, is there a reason to separate these into two functions?

Comment (omerap12) [reply vertical-pod-autoscaler/common/flags.go:51]: I thought it would be a good idea to separate the logging flags (used by klog) from the controller logic flags, but I can merge them if it seems unnecessary.

File: vertical-pod-autoscaler/common/flags.go (hunk 10 above)
Comment (omerap12) [reply vertical-pod-autoscaler/common/flags.go:52]: Those flags should always be exist I think, but It's good to add a check I agree. Fixed :)

**Comment by omerap12** [comment]
Hey @voelzmo , PTAL :)

**Comment by omerap12** [comment]
Merged master branch, PTAL @adrianmoisey

**Review by kwiesmueller — APPROVED** [review]
/lgtm
/approve
